src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.ProjectState.cs

```diff
@@ -292,87 +286,78 @@ public bool OnProjectRemoved(ProjectId id)
                 }
 
                 return builder.ToResult();
             }
 
-            private async Task<DiagnosticAnalysisResult> LoadInitialAnalysisDataAsync(IPersistentStorageService persistentService, TextDocument document, CancellationToken cancellationToken)
+            private async Task<DiagnosticAnalysisResult> LoadInitialAnalysisDataAsync(TextDocument document, CancellationToken cancellationToken)
             {
                 // loading data can be cancelled any time.
                 var project = document.Project;
 
                 var version = await GetDiagnosticVersionAsync(project, cancellationToken).ConfigureAwait(false);
-                var serializer = new DiagnosticDataSerializer(_owner.AnalyzerVersion, version);
+                var serializerVersion = version;
                 var builder = new Builder(project, version);
 
-                if (!await TryDeserializeDocumentDiagnosticsAsync(persistentService, serializer, document, builder, cancellationToken).ConfigureAwait(false))
+                if (!TryGetDiagnosticsFromInMemoryStorage(serializerVersion, document, builder))
                 {
                     return DiagnosticAnalysisResult.CreateEmpty(project.Id, VersionStamp.Default);
                 }
 
                 return builder.ToResult();
             }
 
-            private async Task<DiagnosticAnalysisResult> LoadInitialProjectAnalysisDataAsync(IPersistentStorageService persistentService, Project project, CancellationToken cancellationToken)
+            private async Task<DiagnosticAnalysisResult> LoadInitialProjectAnalysisDataAsync(Project project, CancellationToken cancellationToken)
             {
                 // loading data can be cancelled any time.
                 var version = await GetDiagnosticVersionAsync(project, cancellationToken).ConfigureAwait(false);
-                var serializer = new DiagnosticDataSerializer(_owner.AnalyzerVersion, version);
+                var serializerVersion = version;
                 var builder = new Builder(project, version);
 
-                if (!await TryDeserializeProjectDiagnosticsAsync(persistentService, serializer, project, builder, cancellationToken).ConfigureAwait(false))
+                if (!TryGetProjectDiagnosticsFromInMemoryStorage(serializerVersion, project, builder))
                 {
                     return DiagnosticAnalysisResult.CreateEmpty(project.Id, VersionStamp.Default);
                 }
 
                 return builder.ToResult();
             }
 
-            private async Task SerializeAsync(IPersistentStorageService? persistentService, DiagnosticDataSerializer serializer, Project project, TextDocument? document, object key, string stateKey, ImmutableArray<DiagnosticData> diagnostics)
+            private void AddToInMemoryStorage(VersionStamp serializerVersion, Project project, TextDocument? document, object key, string stateKey, ImmutableArray<DiagnosticData> diagnostics)
```

Comment:
💭 I wouldn't have hated to see this stay asynchronous. In the event we did need to offload this data, we would have the option of implementing it on the original asynchronous paths using temporary storage service instead of persistent storage service. We can still do that, but the change would involve reverting a decent portion of this PR.

Reply:
@sharwell .  That is reasonable.  I'll make this all async again (probably ValueTask).  We can then easily tweak in the future if appropriate.